### PR TITLE
Podcast Player: Refactor front-end icons

### PR DESCRIPTION
**Pull request by WunderBart**
#### Changes proposed in this Pull Request:

Brings back the Player's icons (React-style) after a refactor done in the https://github.com/Automattic/jetpack/pull/15072.

#### Testing instructions:

* Add a Podcast Player block to your page
* Initialize the Player (you can use this URL: https://anchor.fm/s/9400d7c/podcast/rss)
* Open the page preview.
* Click one of the episodes to play it - a loudspeaker icon with text color fill should appear
* Click again the same episode to pause it - the icon should disappear
* Block one of the episode audio URLs to test the error icon
* Click the episode you just blocked - a red exclamation mark icon should appear.
* Click any other episode - the icon should disappear as the error message does.

#### Example screenshots:

| Episode Playing  | Episode Error |
| ------------- | ------------- |
| Screenshot 2020-03-24 17 25 38  | Screenshot 2020-03-24 17 25 23  |

#### Proposed changelog entry

* Podcast Player: Refactor front-end icons react-style

### CODE REVIEW

**Review by marekhrabe — DISMISSED** [review]
Looks good to me!

**Comment by marekhrabe** [comment]
Pinging @jeryj as he mentioned something in my React PR about icons. Just to make sure the loop was closed.

File: extensions/blocks/podcast-player/icons/episode-error-icon.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+/**
+ * External dependencies
+ */
+import { Path, SVG } from '@wordpress/components';
+
+export default (
+	<SVG height="24" viewBox="0 0 24 24" width="24" xmlns="http://www.w3.org/2000/svg">
```
Comment (jeryj) [inline extensions/blocks/podcast-player/icons/episode-error-icon.js]: Add `role="img"` and an `aria-labelledby="jetpack-podcast-player__error-icon"` to the `<SVG` line, and directly beneath, add `<title id="jetpack-podcast-player__error-icon">{ __( 'Error' ) }</title>` to have "Error" announced for this icon with screen readers.

File: extensions/blocks/podcast-player/icons/episode-playing-icon.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+/**
+ * External dependencies
+ */
+import { Path, SVG } from '@wordpress/components';
+
+export default (
+	<SVG height="24" viewBox="0 0 24 24" width="24" xmlns="http://www.w3.org/2000/svg">
```
Comment (jeryj) [inline extensions/blocks/podcast-player/icons/episode-playing-icon.js]: Add `role="img"` and an `aria-labelledby="jetpack-podcast-player__playing-icon"` to the `<SVG` line, and directly beneath, add `<title id="jetpack-podcast-player__playingr-icon">{ __( 'Playing' ) }</title>` to have "Playing" announced for this icon with screen readers.

File: extensions/blocks/podcast-player/icons/episode-error-icon.js (hunk 1 above)
Comment (jeryj) [reply extensions/blocks/podcast-player/icons/episode-error-icon.js]: And add the `__` dependency.

**Review by jeryj — DISMISSED** [review]
I left a few accessibility pointers here. You can see where I got my implementation suggestions from https://css-tricks.com/accessible-svgs/ and scroll down to the `2. Inline SVG` heading.

File: extensions/blocks/podcast-player/icons/episode-playing-icon.js (hunk 2 above)
Comment (WunderBart) [reply extensions/blocks/podcast-player/icons/episode-playing-icon.js]: I noticed that there's an `aria-hidden="true"` added by default to the `SVG` component: 

image

Does it conflict with the `role="img" aria-labelledby="..."` attributes we've just added?

**Comment by marekhrabe** [comment]
This is working well for me in both editor and the frontend. I have one more request for this PR - can we get rid of the complicated conditional fill setting via CSS? 

I propose to have a dedicated class for the playing icon and one for the error icon and apply the `fill` on those, instead of combining episode active flag with player state. This should simplify the CSS and make it less reliant on the surrounding classes.

**Review by marekhrabe — DISMISSED** [review]
Thanks for updating that! I think we are ready

File: extensions/blocks/podcast-player/icons/episode-playing-icon.js (hunk 2 above)
Comment (jeryj) [reply extensions/blocks/podcast-player/icons/episode-playing-icon.js]: Hmm... Yeah, because `aria-hidden` indicates to the accessibility API that this should be ignored. Does setting `aria-hidden={ undefined }` on the `<SVG` work to remove it?

Comment (WunderBart) [reply extensions/blocks/podcast-player/icons/episode-playing-icon.js]: Unfortunately, that attribute is hard-coded https://github.com/WordPress/gutenberg/blob/master/packages/primitives/src/svg/index.js#L29 :/

Comment (jeryj) [reply extensions/blocks/podcast-player/icons/episode-playing-icon.js]: Hrmmm.... I wonder what the intention is there. I guess the alternative is to just use `<VisuallyHidden>` text.

File: extensions/blocks/podcast-player/components/playlist.js [hunk 3]
```diff
@@ -8,7 +8,31 @@ import { __ } from '@wordpress/i18n';
...
+	const icon = episodeIcons[ name ];
+
+	if ( ! icon ) {
+		// Return empty element - we need it for layout purposes.
+		return <span className={ className } aria-hidden="true" />;
+	}
+
+	return (
+		<span className={ `${ className } ${ className }--${ name }` } aria-hidden="false">
+			{ icon }
```
Comment (jeryj) [inline extensions/blocks/podcast-player/components/playlist.js]: We can insert the previous icon text as a `<VisuallyHidden>` text here, since the `icon` has `aria-hidden` hardcoded as `true`. So, `<VisuallyHidden>Playing</VisuallyHidden>`, etc.

File: extensions/blocks/podcast-player/components/playlist.js [hunk 4]
```diff
@@ -8,7 +8,31 @@ import { __ } from '@wordpress/i18n';
...
+
+	const icon = episodeIcons[ name ];
+
+	if ( ! icon ) {
+		// Return empty element - we need it for layout purposes.
+		return <span className={ className } aria-hidden="true" />;
+	}
+
+	return (
+		<span className={ `${ className } ${ className }--${ name }` } aria-hidden="false">
```
Comment (jeryj) [inline extensions/blocks/podcast-player/components/playlist.js]: We likely don't need the `aria-hidden="false"` here.

File: extensions/blocks/podcast-player/components/playlist.js [hunk 5]
```diff
@@ -8,7 +8,31 @@ import { __ } from '@wordpress/i18n';
...
+		name = 'error';
+	} else if ( isPlaying ) {
+		name = 'playing';
+	}
+
+	const icon = episodeIcons[ name ];
+
+	if ( ! icon ) {
+		// Return empty element - we need it for layout purposes.
+		return <span className={ className } aria-hidden="true" />;
```
Comment (jeryj) [inline extensions/blocks/podcast-player/components/playlist.js]: Since there's no content in this element that would be announced via a screen reader, I think you can safely remove the `aria-hidden` attribute.

File: extensions/blocks/podcast-player/icons/episode-icons.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+			aria-labelledby={ id }
+		>
+			<title id={ id }>{ __( title, 'jetpack' ) }</title>
+			{ children }
+		</SVG>
+	);
+};
+
+export const playing = (
+	<EpisodeIcon name="playing" title="Playing">
```
Comment (jeherve) [inline extensions/blocks/podcast-player/icons/episode-icons.js]: Let's translate that title, and the Error one below.

**Review by jeherve — CHANGES_REQUESTED** [review]
This tests well for me, including in IE11. I only have a minor remark.

File: extensions/blocks/podcast-player/components/playlist.js (hunk 3 above)
Comment (marekhrabe) [reply extensions/blocks/podcast-player/components/playlist.js]: `VisuallyHidden` is a Gutenberg component. Bringing it in would likely mean loading all Gutenberg components on the frontend too. Let's refrain from it, unless 100% necessary.

Comment (jeryj) [reply extensions/blocks/podcast-player/components/playlist.js]: Ah, we can just do a `<span className="whatever-the-visually-hidden-class-floating-around-is">`. If there isn't a global one we can add it to our block. https://a11yproject.com/posts/how-to-hide-content/

File: extensions/blocks/podcast-player/icons/episode-icons.js (hunk 6 above)
Comment (WunderBart) [reply extensions/blocks/podcast-player/icons/episode-icons.js]: The translation is called from the `EpisodeIcon` component: https://github.com/Automattic/jetpack/blob/16b41b13007b4db77d903d4606a57a3850ad5d7a/extensions/blocks/podcast-player/icons/episode-icons.js#L19

Does that look OK?

File: extensions/blocks/podcast-player/components/playlist.js (hunk 5 above)
Comment (WunderBart) [reply extensions/blocks/podcast-player/components/playlist.js]: This was required by a review here https://github.com/Automattic/jetpack/pull/15072#discussion_r397186162. Calling @getdave on this one.

File: extensions/blocks/podcast-player/icons/episode-icons.js (hunk 6 above)
Comment (jeherve) [reply extensions/blocks/podcast-player/icons/episode-icons.js]: That won't work I am afraid. `__()` will not accept variables, only strings, so you'll need to have the translation happening here.

Comment (marekhrabe) [reply extensions/blocks/podcast-player/icons/episode-icons.js]: I think I see what @jeherve meant. Afaik, `__` calls are statically extracted and that's how they end up in the list of strings to be translated (at least that's how we do that in other projects in A8C). 

Because of that reason, `__` always needs to have the string directly as an argument. Not because of functionality (JS would obviously pass the value from variable) but because we use static analysis on code that looks for `__` and extracts string.

tldr: use prop directly in `<title>` (no translation) and translate it where you call `EpisodeIcon`, eg: `title={ __( 'Playing', 'jetpack' ) }`

File: extensions/blocks/podcast-player/components/playlist.js (hunk 5 above)
Comment (jeryj) [reply extensions/blocks/podcast-player/components/playlist.js]: @getdave I don't see a semantic reason for adding this. There's no content to hide in this instance. Also, as a rule of thumb, we should only use `aria-hidden` where the content will be hidden for _all users_. The time to use it is when we will be providing an alternative content or the element will read duplicate, unnecessary content. https://www.w3.org/TR/wai-aria-1.1/#aria-hidden

File: extensions/blocks/podcast-player/icons/episode-icons.js (hunk 6 above)
Comment (WunderBart) [reply extensions/blocks/podcast-player/icons/episode-icons.js]: Fixed in https://github.com/Automattic/jetpack/pull/15101/commits/57dcf7cc5454d925c1865c5af2a5b9f3f955d4a2

File: extensions/blocks/podcast-player/components/playlist.js (hunk 4 above)
Comment (WunderBart) [reply extensions/blocks/podcast-player/components/playlist.js]: Removed in https://github.com/Automattic/jetpack/pull/15101/commits/a1dc70a4ad681d7c36639a38a84f06c1cf463dd2

File: extensions/blocks/podcast-player/icons/episode-playing-icon.js (hunk 2 above)
Comment (WunderBart) [reply extensions/blocks/podcast-player/icons/episode-playing-icon.js]: Updated in https://github.com/Automattic/jetpack/pull/15101/commits/bd45c78466addc8bd5b59474201b6b7cbe1dfbfb.

Opened a PR in Gutenberg for custom SVG `aria-hidden` support: https://github.com/WordPress/gutenberg/pull/21173

As agreed in p1585170298230600-slack-ajax, we will not make any further changes in this matter here and will follow up once the Gutenberg PR gets reviewed & merged/closed.

File: extensions/blocks/podcast-player/components/playlist.js (hunk 3 above)
Comment (WunderBart) [reply extensions/blocks/podcast-player/components/playlist.js]: See https://github.com/Automattic/jetpack/pull/15101#discussion_r398618668

File: extensions/blocks/podcast-player/components/playlist.js (hunk 5 above)
Comment (getdave) [reply extensions/blocks/podcast-player/components/playlist.js]: I see what you're getting at. However.

> Authors MAY, with caution, use aria-hidden to hide visibly rendered content from assistive technologies only if the act of hiding this content is intended to improve the experience for users of assistive technologies by removing **redundant or extraneous content.**

A span which has no content in it which is just sitting in the DOM for no reason is to me redundant. You only have it there as a placeholder for a logo which doesn't exist.

It's pretty minor though so feel free to do otherwise if you feel that's best.

Comment (jeryj) [reply extensions/blocks/podcast-player/components/playlist.js]: I think the confusion here is what they mean by "content." They're referring it to  it as "visibly rendered content." As in, something displayed on the screen. An empty span has no "visibly rendered content," so there's no visibly rendered content to remove.

Comment (WunderBart) [reply extensions/blocks/podcast-player/components/playlist.js]: Removed in https://github.com/Automattic/jetpack/pull/15101/commits/1c60d16250f0df3c681ab2607ad38decb40c4200

**Review by jeherve — DISMISSED** [review]
This looks good to me. It should be good to merge.

**Review by jeherve — APPROVED** [review]

**Comment by marekhrabe** [comment]
r204964-wpcom
